proofs/Settlement.thy

```diff
@@ -0,0 +1,160 @@
+theory Settlement
+
+imports Main
+
+begin
+
+text "This file contains an analysis of the settlement algorithm in the TokenNetwork contract.
+https://raiden-network-specification.readthedocs.io/en/latest/smart_contracts.html#protocol-values-and-settlement-algorithm-analysis
+The result so far confirms two things:
+  lemma s1_correct: The value calculated as 'S1 = RmaxP1 - SL2' is equal to 'S1 = D1 - W1 + T2 - T1 - L1'.
+  lemma s2_correct: Similarly for 'S2 = RmaxP2 - SL1' and 'S2 = D2 - W2 + T1 - T2 - L2'.
+The required conditions appear in the statements of the lemma.
+"
+
+text "TODO:
+* Make sure that, you can only lose tokens if you submit an older balance proof.
+"
+
+type_synonym impl_number = "int option"
+
+definition valid :: "int \<Rightarrow> bool"
+  where "valid a = (0 \<le> a \<and> a < 32)"
```

Comment:
You're right.

It's `[0, 32)` only because Isabelle finds counter-examples faster with the smaller region (but that's relevant only when the proof fails).

I'll try to change this into `[0, 2**256)` and see if the proofs still go through.

Reply:
When I replace 32 with 2**256, the proof about S1 goes through but S2 fails.  Investigating...

Reply:
Now I need `(D1 + D2) < 2 ** 256`.  Not sure why.

Reply:
Done changing into "2 ** 256*.

> question, why do you use 0 \<le> a on the right, but a < 32 on the left?

Because 0 is a `uint256` but `2 ^ 256` is no longer a `uint256`.